src/JuliusSweetland.OptiKey.Core/UI/ViewModels/MainViewModel.ServiceEventHandlers.cs

```diff
@@ -2372,46 +2372,152 @@ private void SelectVoice(string voice)
+            {
+                //handle the commands
+                StepListCommands(TimerList.First().SingleKeyValue, TimerList.First().MultiKeySelection, TimerList.First().CommandList);
+                //remove the commands
+                TimerList.RemoveAt(0);
+            }            
+        }
+        public class Commands
+        {
+            public KeyValue SingleKeyValue { get; set; }
+            public List<string> MultiKeySelection { get; set; }
+            public List<string> CommandList { get; set; }
+        }
+        public List<Commands> TimerList = new List<Commands>();
+
+        private void StepList(KeyValue singleKeyValue, List<string> multiKeySelection)
         {
-            if (!string.IsNullOrEmpty(command))
+            if (!string.IsNullOrEmpty(singleKeyValue.String))
             {
-                KeyValue singleKeyValue;
-                string[] stringSeparators = new string[] { "<" };
-                foreach (var vStep in command.Split(stringSeparators, StringSplitOptions.None).ToList())
+                Log.InfoFormat("StepList called with singleKeyValue [{0}]", singleKeyValue.String);
+
+                List<string> commandList = singleKeyValue.String.Split('<').ToList();
+                if (commandList.Count > 0)
                 {
-                    Log.DebugFormat("Performing StepList step: {0}.", vStep);
-                    if (vStep.StartsWith("Action>") &&
-                        (Enum.TryParse<FunctionKeys>(vStep.Substring(7), out FunctionKeys result)))
+                    //old but works:
+                    if (false)
+                        StepListCommands(singleKeyValue, multiKeySelection, commandList);
+                    //new but not ready:
+                    else
                     {
-                        singleKeyValue = new KeyValue(result);
-                        HandleFunctionKeySelectionResult(singleKeyValue);
+                        Commands cmd = new Commands
+                        {
+                            SingleKeyValue = singleKeyValue,
+                            MultiKeySelection = multiKeySelection,
+                            CommandList = commandList
+                        };
+                        //Create a new timer to handle the loop async
+                        TimerStepList(cmd);
                     }
-                    else if (vStep.StartsWith("Keyboard>") || vStep.StartsWith("KeyboardAndReturn>"))
+                }
+            }
+        }
+
+        private void StepListCommands(KeyValue singleKeyValue, List<string> multiKeySelection, List<string> commandList)
+        {
+            Log.InfoFormat("StepListCommands called with a list of {0} commands", commandList.Count);
+
+            KeyValue keyValue;
+            while (commandList.Count > 0)
+            {
+                Log.InfoFormat("Performing StepList step: {0}.", commandList.First());
+
+                if (commandList.First().StartsWith("Loop>"))
```

Comment:
Fixed some code:
>if (commandList.First().StartsWith("Loop>"))
                {
                    var count = Int32.Parse(commandList.First().Substring(5));
                    //if this loop has a nested loop then set the end of the loopList to that of the last end loop command
                    var vEnd = (commandList.FindIndex(1, x => x.StartsWith("Loop>")) > 0 
                        && commandList.FindIndex(1, x => x.StartsWith("Loop>")) < commandList.IndexOf("/Loop>"))
                        ? commandList.LastIndexOf("/Loop>") - 1
                        : commandList.IndexOf("/Loop>") - 1;                    
                    for (int i = 0; i < count; i++)
                    {
                        var loopList = commandList.GetRange(1, vEnd);
                        StepListCommands(singleKeyValue, multiKeySelection, loopList);
                    }
                    commandList.RemoveRange(0, vEnd);
                }